cmd/kubeops/internal/handler/dashboardcompat.go

```diff
@@ -0,0 +1,121 @@
+	h3chart "helm.sh/helm/v3/pkg/chart"
+	h3 "helm.sh/helm/v3/pkg/release"
+	h2chart "k8s.io/helm/pkg/proto/hapi/chart"
+	h2 "k8s.io/helm/pkg/proto/hapi/release"
+)
+
+// generatedYamlHeader is prepended to YAML generated from the internal map[string]interface{} representation.
+const generatedYamlHeader = "# Not original YAML! Generated from parsed representation."
+
+type dashboardCompatibleRelease struct {
+	Name      string                    `json:"name,omitempty"`
+	Info      h2.Info                   `json:"info,omitempty"`
+	Chart     dashboardCompatibleChart  `json:"chart,omitempty"`
+	Config    dashboardCompatibleConfig `json:"config,omitempty"`
+	Manifest  string                    `json:"manifest,omitempty"`
+	Version   int                       `json:"version,omitempty"`
+	Namespace string                    `json:"namespace,omitempty"`
+}
+
+type dashboardCompatibleChart struct {
+	Files     []*any.Any                `json:"files,omitempty"`
+	Metadata  h3chart.Metadata          `json:"metadata,omitempty"`
+	Templates []*h2chart.Template       `json:"templates,omitempty"`
+	Values    dashboardCompatibleValues `json:"values,omitempty"`
+}
+
+type dashboardCompatibleValues struct {
+	Raw string `json:"raw,omitempty"`
+}
+
+type dashboardCompatibleConfig struct {
+	Raw string `json:"raw,omitempty"`
+}
+
+func newDashboardCompatibleRelease(h3r h3.Release) dashboardCompatibleRelease {
+	return dashboardCompatibleRelease{
+		Name:      h3r.Name,
+		Info:      h2.Info{Status: compatibleStatus(*h3r.Info)},
+		Chart:     compatibleChart(*h3r.Chart),
+		Config:    compatibleConfig(h3r),
+		Manifest:  h3r.Manifest,
+		Version:   h3r.Version,
+		Namespace: h3r.Namespace,
+	}
+}
+
+func compatibleChart(h3c h3chart.Chart) dashboardCompatibleChart {
+	return dashboardCompatibleChart{
+		Files:     compatibleFiles(h3c.Files),
+		Metadata:  *h3c.Metadata,
+		Templates: compatibleTemplates(h3c.Templates),
+		Values:    compatibleValues(h3c),
+	}
+}
+
+func compatibleFiles(h3files []*h3chart.File) []*any.Any {
+	anys := make([]*any.Any, len(h3files))
+	for i, f := range h3files {
+		anys[i] = &any.Any{
+			TypeUrl: f.Name,
+			Value:   f.Data,
+		}
+	}
+	return anys
+}
+
+func compatibleTemplates(h3templates []*h3chart.File) []*h2chart.Template {
+	templates := make([]*h2chart.Template, len(h3templates))
+	for i, t := range h3templates {
+		templates[i] = &h2chart.Template{
+			Name: t.Name,
+			Data: t.Data,
+		}
+	}
+	return templates
+}
+
+func compatibleValues(h3c h3chart.Chart) dashboardCompatibleValues {
+	return dashboardCompatibleValues{
+		Raw: valuesToYaml(h3c.Values),
+	}
+}
+
+func compatibleConfig(h3r h3.Release) dashboardCompatibleConfig {
+	return dashboardCompatibleConfig{
+		Raw: valuesToYaml(h3r.Config),
+	}
+}
+
+// valuesToYaml serializes to YAML and prepends an informative header.
+// It assumes that the serialization succeeds.
+func valuesToYaml(values map[string]interface{}) string {
+	marshaled, _ := yaml.Marshal(values)
```

Comment:
Why do you want to ignore the error rather than handling it here? This could result in hard-to-discover bugs later (ie. getting an empty values without any known reason why, I think?)

Actually, this looks like another place where we could just use the helm2 pkg? See

https://github.com/helm/helm/blob/bbdfe5e7803a12bbdf97e94cd847859890cf4050/pkg/tiller/release_server.go#L157

where `Raw` is set using https://github.com/helm/helm/blob/release-2.16/pkg/chartutil/values.go#L45

Reply:
I thought there would never occur any error there, and abandoning that assumption would contaminate the entire file with `error` and `nil` everywhere without adding any real value (or so I thought).

There _should_ exist a total function from `h3.Release` to some Dashboard-compatible type, shouldn't there? (`newDashboardCompatibleRelease` is my attempt at constructing such a function.) Maybe @andresmgot's suggestion of storing the YAML separately is the best solution; then we wouldn't need to marshal anything at all.

Finally, I don't see how [`YAML`](https://github.com/helm/helm/blob/release-2.16/pkg/chartutil/values.go#L45) could be used here. It works only on members of `Values` from Helm 2, but we have a `map[string]interface{}`.

Reply:
In the unlikely case that the `map[string]interface{}` couldn't be marshaled, returning an error here would be helpful to identify the issue. I would either log the error (if you want this function to succeed in any case, even if the values cannot be marshalled) or return it and handle it in an upper level.

Reply:
I'd suggest one of these options for now, to keep the code simple:
* Panic if serialization failed.
* Ignore any serialization error (current solution).

The rationale would be the same no matter which one we choose: This won't hit production anyway (because we need to figure out how to store/access `values.yaml`), and it's good enough™ for us to be able to keep working on Helm 3 support.

Reply:
why don't you log the error instead? A `panic` can be disruptive in the whole process and you don't need to return the error and obscure the code if you just log the error.

Reply:
Sounds reasonable. I'll log it instead.